### ISSUE TITLE
Informational ISUPPORT or cap for joined channel restoration

### ISSUE THREAD

**Issue by emersion** [ircv3/ircv3-ideas#91]
When a client reconnects to a bouncer or bouncer-like IRC server such as soju, znc, or ergo, it's not desirable to re-join all previously joined channels: the server is responsible for keeping track of joined channels and will send `JOIN` messages to the client.

However clients have no way to discover that a server will keep track of joined channels. This leads to annoying behavior:

- User connects with client A
- User joins #ircv3
- Client A gets disconnected (closed or offline)
- User connects with client B (e.g. different device)
- User leaves #ircv3
- Client A reconnects (user launches it again, or the device becomes online)
- Client A sends `JOIN #ircv3` because it was previously opened
- User is annoyed because #ircv3 shows up again, even if they left the channel

It would be nice to have an ISUPPORT token to stop clients from sending `JOIN` after connecting, e.g. `JOINRESTORE`.

cc @slingamn @DarthGandalf

**Comment by DarthGandalf** [comment]
If user wants to leave a channel, he shouldn't have it in the autojoin list in any client.

I think silently not joining the channels even though they are configured in the client will be more confusing

**Comment by emersion** [comment]
The idea here is to have the client behave correctly by default, without requiring the user to turn off the autojoin settings. BTW, some clients don't even provide autojoin settings (e.g. WeeChat or all of my clients).

If the user has configured an explicit autojoin setting in the client, it should be used. Otherwise, the client can use this hint from the server.

**Comment by DarthGandalf** [comment]
> Otherwise, the client can use this hint from the server.

If client isn't going to JOIN any channel, this token wouldn't be needed, no?

In which case the client tries to join a channel without being told to join the channel? Does weechat somehow remember what it joined before without explicit setting?

**Comment by flashcode** [comment]
Oh, I just added a way in WeeChat a way to automatically save joined channels (with server option "autojoin_dynamic").

Anyway, this can be turned of per server, in case in future WeeChat implements such tracking provided by the server.

@emersion: what do you mean by no "autojoin" in WeeChat? There's such option since the beginning, it was just set manually by the user until recently.

**Comment by emersion** [comment]
> Does weechat somehow remember what it joined before without explicit setting?

Yes. WeeChat will re-join channels on reconnect, but has no way to turn this feature off.

Even if WeeChat had a way to turn this off, it would be better to just Do The Right Thing without any need for manual configuration.

**Comment by emersion** [comment]
@flashcode hmmm, when I `/reconnect -all` in WeeChat, it sends `JOIN` for all channels, and I found no way to turn that off. But maybe I missed something?

**Comment by flashcode** [comment]
@emersion: you can do `/reconnect -all -nojoin` (by default all channels are joined)

**Comment by emersion** [comment]
It would also be useful to have a batch wrapping the JOIN burst sent during connection registration. This way, clients can know when the server is done pushing the initial channel state. This is useful when a clients is opened with an `irc://` URI for instance, and needs to figure out whether it should prompt the user to join a channel.

**Comment by slingamn** [comment]
That sounds reasonable, especially if we had an explicit CAP to enable it?

**Comment by slingamn** [comment]
#503 was a more elaborate response to the original issue.

**Comment by jwheare** [comment]
https://github.com/ircv3/ircv3-specifications/pull/503

**Comment by emersion** [comment]
I have no plans to support the persistence IRC extension, but I still want to address the problems outlined in this issue.

**Comment by slingamn** [comment]
Being able to unambiguously identify the end of the join burst seems basically orthogonal to solving the autojoin+bouncer interaction. But batching the join burst seems fine; if there was an explicit cap for batching it, I would implement it.
